File: Java/P4/Civitas/src/civitas/tituloPropiedad.java
/*
 * To change this license header, choose License Headers in Project Properties.
 * To change this template file, choose Tools | Templates
 * and open the template in the editor.
 */
package civitas;

/**
 *
 * @author arturo
 */
public class tituloPropiedad {
   private boolean hipotecado;
   private int numCasas;
   private int numHoteles;
   static private final float factorInteresesHipoteca=(float) 1.1;
   private float alquilerbase;
   private float factor_rev;
   private float precioBaseHipoteca;
   private float precio_compra;
   private float precio_edif;
   private String nombre;
   // sucede y nosucede se utilizan para comprobar 
   private String sucede="Operacion realizada \n"; 
   private String nosucede="Operacion no realizada \n";
   private float precioalquiler;
   private Jugador propietario;
   private int ip;
    
   public tituloPropiedad(String nombre, float alquilerbase, float factor_rev, float precioBaseHipoteca, float precio_compra, float precio_edif){       
      this.nombre=nombre;
      this.factor_rev=factor_rev;
      this.precioBaseHipoteca=precioBaseHipoteca;
      this.precio_compra=precio_compra;
      this.precio_edif=precio_edif;
      this.numCasas=0;
      this.numHoteles=0;
      this.hipotecado=false;
      this.alquilerbase=alquilerbase;
      propietario=null;
   }
   
   int getIp(){
       return ip;
   }
  
   @Override //toString es un método de la clase String, por lo que esto indica que no me estoy equivocando al tito netbeans 
   public String toString(){
       return "*TituloPropiedad*" + "\n\t Nombre: " + nombre + "\n\t PrecioCompra: " + precio_compra +
                "\n\t AlquilerBase: " + alquilerbase + "\n\t HipotecaBase: " + precioBaseHipoteca + 
                "\n\t PrecioEdificar: " + precio_edif + "\n\t FactorRevalorizacion: " + factor_rev +
                "\n\t NumHoteles: " + numHoteles + "\n\t NumCasas: " + numCasas + 
                "\n\t Hipotecada: " + getHipotecado() + "\n"; 
   }
   
   public float getPrecioAlquiler(){   
       if(!hipotecado && this.propietarioEncarcelado()==false){
           this.precioalquiler=(float) (alquilerbase*(1+(numCasas*0.5)+(numHoteles*2.5)));  
       }else{
           this.precioalquiler=0;
       }
       
       return precioalquiler;
   }
   
   public Jugador getPropietario(){
       return propietario;
   }
   
   float getImporteCancelarHipoteca(){
       float aux;      
       aux=precioBaseHipoteca*factorInteresesHipoteca;
       return aux;
   }
   
   public boolean getHipotecado(){
       return hipotecado;
   }
   
   public String getNombre(){
       return nombre;
   }
   
   public int getNumCasas(){
       return numCasas;
   }
   
   public int getNumHoteles(){
       return numHoteles;
   }
   
   float getPrecioCompra(){
       return precio_compra;
   }
   
   private boolean esEsteElPropietario(Jugador jugador){
       boolean aux=false;
       
       if(propietario==jugador){
           aux=true;
       }
       
       return aux;
   }
   
   boolean cancelarHipoteca(Jugador jugador){
       boolean aux=false;
       
       if(hipotecado==true && esEsteElPropietario(jugador)){
           propietario.paga(getImporteCancelarHipoteca());
           hipotecado=false;
           aux=true;
       }
       
       return aux;
   }
   
   boolean hipotecar(Jugador jugador){
       if(hipotecado==false && esEsteElPropietario(jugador)){
           propietario.recibe(precioBaseHipoteca);
           hipotecado=true;
       }
       
       return hipotecado;
   }
   
   boolean tienePropietario(){
       boolean aux=false;
       
       if(propietario!=null){
            aux=true; 
       }
       
       return aux;
    }
   
   void tramitarAlquiler(Jugador jugador){
       if(this.tienePropietario()==true && !esEsteElPropietario(jugador)){
           jugador.pagaAlquiler(getPrecioAlquiler());
           propietario.recibe(getPrecioAlquiler());
           System.out.print(sucede);
       }else{
           System.out.print(nosucede);
       }
   }
     
   private boolean propietarioEncarcelado(){
       boolean aux=false;
       
       if(!tienePropietario() || propietario.encarcelado){
           aux=true;
       }
       
       return aux;
   }
   
   int cantidadCasaHoteles(){
       return numCasas+numHoteles;
   }
   
   boolean derruirCasas(int n, Jugador jugador){
       boolean aux=false;
       
       if(esEsteElPropietario(jugador) && numCasas>=n){
            numCasas=numCasas-n;
            aux=true;
           if(aux){
               System.out.print(sucede);
           }else{
               aux=false;
               System.out.print(nosucede);  
           }
       }
       return aux;
   }
   
   boolean derruirHoteles(int n, Jugador jugador){
       boolean aux=false;
       
       if(esEsteElPropietario(jugador) && numHoteles>=n){
            numHoteles=numHoteles-n;
            aux=true;
           if(aux){
               System.out.print(sucede);
           }else{
               aux=false;
               System.out.print(nosucede);
           }
       }
       return aux;
   }
   
   float getPrecioVenta(){
       return precio_compra+precio_edif*cantidadCasaHoteles()*factor_rev;
   }
   
   public float getPrecioEdificar(){
       return precio_edif;
   }
   
   boolean construirCasa(Jugador jugador){
       boolean aux=false;
       
       if(this.esEsteElPropietario(jugador)){
           propietario.paga(precio_edif);
           numCasas++;
           aux=true;
       }
       
       return aux;
   }
   
   boolean construirHotel(Jugador jugador){
       boolean aux=false;
       
       if(esEsteElPropietario(jugador)==true && numCasas>=4){
           propietario.paga(precio_edif);
           numHoteles++;
           derruirCasas(4, jugador);
           aux=true;
       }
       
       return aux;
   }
   
   boolean Comprar(Jugador jugador){
       boolean aux=false;
       
       if(!tienePropietario()){
           aux=true;
           propietario=jugador;
           jugador.paga(getPrecioVenta());
       }
       
       return aux;
   }
   
   void actualizaPropietarioPorConversion(Jugador jugador){
       this.propietario=jugador;
   }
   
   boolean Vender(Jugador jugador){
       boolean aux=false;
       
       if(hipotecado==false && esEsteElPropietario(jugador)==true){
           System.out.print(getPrecioVenta()+" es lo que recibe por vender\n");
           jugador.recibe(getPrecioVenta());
           derruirCasas(numCasas, jugador);
           derruirHoteles(numHoteles, jugador);
           actualizaPropietarioPorConversion(null);
           aux=true;
       }
       
       return aux;
   }
}

File: Ruby/Civitas/lib/titulo_propiedad.rb
# To change this license header, choose License Headers in Project Properties.
# To change this template file, choose Tools | Templates
# and open the template in the editor.

# encoding: utf-8
require_relative "casilla"
require_relative "diario"
require_relative "jugador"

module Civitas
  class TituloPropiedad
    @@Factor_intereses_hipoteca=1.1
    
    def initialize(nombre, base, factor, hipoteca, compra, edificar)
      @nombre=nombre
      @alquiler_base=base
      @factor_revalorizacion=factor
      @hipoteca_base=hipoteca
      @precio_compra=compra
      @precio_edificar=edificar
      @num_casas=0
      @num_casas=0
      @num_hoteles=0
      @hipotecado=false
      @propietario
    end
    
    attr_reader   :nombre, :alquiler_base, :factor_revalorizacion, 
                  :hipoteca_base, :precio_compra, :precio_edificar
    attr_accessor :num_casas, :num_hoteles, :hipotecado, :propietario
    
    def actualizar_propietario_por_conversion(jugador)
      @propietario=jugador
    end
    
    def get_precio_alquiler
      if(hipotecdo==true || propietario_encarcelado)
        precio=0
      else
        precio=@alquiler_base*(1+@num_casas*0.5+@num_hoteles*2.5)
      end
      
      return precio
    end
    
    def get_importe_cancelar_hipoteca
      importe=@hipoteca_base*@@Factor_intereses_hipoteca
      
      return importe
    end
    
    def tramitar_alquiler(jugador)
      actualizar_propietario_por_conversion(jugador)
      if(propietario!=nil && propietario!=jugador)
        jugador.paga(get_precio_alquiler)
        @propietario.recibe(get_precio_alquiler)
      end
    end
    
    def propietario_encarcelado
      p_encarcelado=false
      
      if(tiene_propietario && propietario.encarcelado==true)
        p_encarcelado=true
      end 
      
      return p_encarcelado
    end
    
    private
    
    def es_este_el_propietario(jugador)
      actualizar_propietario_por_conversion(jugador)
      es=false
      
      if(propietario==jugador)
        es=true
      end
      
      return es
    end
    
    def get_importe_hipoteca
      importe=(@hipoteca_base*(1+(@num_casas*0.5)+(@num_hoteles*2.5)))
        
      return importe 
    end
    
    public 
    
    def get_hipotecado
      esta_hipotecado=@hipotecado
      
      return esta_hipotecado
    end
    
    def get_importe_cancelar_hipoteca
      importe_hipoteca=@hipoteca_base*@@Factor_intereses_hipoteca
      
      return importe_hipoteca
    end
    
    def get_precio_compra
      precio=@precio_compra
      
      return precio
    end
    
    def get_propietario
      p=Jugador.new(" ")
      p=@propietario
      
      return p
    end
    
    private
    
    def get_precio_venta
      venta=@precio_compra+@precio_edificar*cantidad_casas_hoteles*@factor_revalorizacion
      
      return venta
    end
    
    def get_precio_alquiler
      precio=@alquiler_base*(1+(@num_casas*0.5)+(@num_hoteles*2.5))
    
      if(hipotecado || propietario_encarcelado)
        precio=0
      end
      
      return precio
    end
    
    public 
    
    def get_precio_edificar
      precio=@precio_edificar
      
      return precio
    end
    
    def get_num_casas
      casas=@num_casas
      
      return casas
    end
    
    def cantidad_casas_hoteles
      cantidad=@num_casas+@num_hoteles
      
      return cantidad
    end
    
    def get_num_hoteles
      hoteles=@num_hoteles
      
      return hoteles
    end
    
    def get_nombre
      nom=@nombre
      
      return nom
    end
    
    def derruir_casas(n, jugador)
      derruidas=false
      
      if(propietario==jugador && num_casas>=n)
        @num_casas=@num_casas-n
        derruidas=true
      end
      
      return derruidas
    end
    
    def derruir_hoteles(n, jugador)
      derruidos=false
      
      if(propietario==jugador && num_hoteles>=n)
        @num_hoteles=@num_hoteles-n
        derruidos=true
      end
      
      return derruidos
    end
    
    def comprar(jugador)
      comprar=false
      
      if(!tiene_propietario)
        actualizar_propietario_por_conversion(jugador)
        comprar=true
        @propietario.paga(get_precio_compra)
      end
      
      return comprar
    end
    
    def vender(jugador)
      vendido=false
      
      if(propietario==jugador && hipotecado==false)
        actualizar_propietario_por_conversion(jugador)
        jugador.recibe(get_precio_venta)
        derruir_casas(num_casas, jugador)
        derruir_hoteles(num_hoteles, jugador)
        @propietario=nil
        vendido=true
      end
      
      return vendido
    end
    
    def hipotecar(jugador)
      hipotecar=false
      
      if(!hipotecado && es_este_el_propietario(jugador))
        actualizar_propietario_por_conversion(jugador)
        @propietario.recibe(get_importe_hipoteca)
        @hipotecado=true
        hipotecar=true
      end
      
      return hipotecar
    end
    
    def cancelar_hipoteca(jugador)
      cancelar=false
      
      if(hipotecado && es_este_el_propietario(jugador))
        actualizar_propietario_por_conversion(jugador)
        @propietario.paga(get_importe_cancelar_hipoteca)
        @hipotecado=true
        cancelar=true
      end
      
      return cancelar
    end
    
    def construir_casa(jugador)
      construir=false
      
      if(es_este_el_propietario(jugador))
        @propietario.paga(get_precio_edificar)
        @num_casas+=1
        construir=true
      end
      
      return construir
    end
    
    def constuir_hotel(jugador)
      construir=false
      
      if(es_este_el_propietario(jugador))
        @propietario.paga(get_precio_edificar)
        @num_hoteles+=1
        construir=true
      end
      
      return construir
    end
    
    private
    
    def propietario_encarcelado
      encarcelado=false
      
      if(propietario.get_encarcelado && tiene_propietario)
        encarcelado=true
      end
      
      return encarcelado
    end
    
    public
    
    def tiene_propietario
      tiene=false
      
      if(propietario!=nil)
        tiene=true
      end
      
      return tiene
    end
    
    def to_s
      puts "\n\t\t-Titulo en Propiedad-
            \n\tNombre: #{@nombre}
            \n\tPrecio de compra: #{@precio_compra}
            \n\tAlquiler base: #{@alquiler_base}
            \n\tHipoteca base: #{@hipoteca_base}
            \n\tPrecio de edificacion: #{@precio_edificar}
            \n\tNombre del propietario: #{@propietario}
            \n\tNumero de casas: #{@num_casas}
            \n\tNumero de hoteles: #{@num_hoteles}
            \n\tHipotecado: #{@hipotecado}"
    end
  end
end
